File: backend/app/schemas/backtest.py
"""Response contract for GET /strategies/{id}/backtest -- the
structured, chat-independent counterpart to BacktestAnalystAgent's
prose. Every number here comes from the existing, unmodified
run_backtest/metrics.py/simulate_buy_and_hold functions; this file
adds no calculation of its own, only typed shape around results those
functions already produce.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class BacktestMetrics(BaseModel):
    model_config = ConfigDict(extra="forbid")

    symbol: str
    # None when there wasn't enough price history in the window to run
    # a single day of simulation -- an honest gap, not a zero.
    period_start: datetime | None
    period_end: datetime | None
    starting_capital: float
    ending_capital: float | None
    total_return_pct: float | None
    max_drawdown_pct: float | None
    sharpe_ratio: float | None
    trade_count: int
    win_rate_pct: float | None


class BacktestComparison(BaseModel):
    model_config = ConfigDict(extra="forbid")

    # The benchmark is always buy-and-hold of this same symbol over
    # the identical window -- deterministic, no entity resolution.
    symbol: str
    total_return_pct: float | None
    max_drawdown_pct: float | None


class BacktestStructuredResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    # What starting_capital actually represents for this run --
    # e.g. "your strategy's declared trading pool" -- resolved via the
    # same shared helper the chat-based Agent uses, so the two paths
    # can never silently disagree on this.
    capital_label: str
    rules: list[BacktestMetrics]
    benchmarks: list[BacktestComparison]
    # Populated when the strategy has more than one rule -- each rule
    # was backtested independently with the FULL assumed balance; this
    # never models the Opportunity Engine's real cross-asset capital
    # competition. See backtest_analyst_agent.py for the identical
    # disclosure used in the chat path.
    limitations: list[str] = []

File: frontend/src/types/backtest.ts
// Mirrors app/schemas/backtest.py exactly -- no calculation happens
// here, only typed shape around numbers the backend already computed.

export interface BacktestMetrics {
  symbol: string;
  period_start: string | null;
  period_end: string | null;
  starting_capital: number;
  ending_capital: number | null;
  total_return_pct: number | null;
  max_drawdown_pct: number | null;
  sharpe_ratio: number | null;
  trade_count: number;
  win_rate_pct: number | null;
}

export interface BacktestComparison {
  symbol: string;
  total_return_pct: number | null;
  max_drawdown_pct: number | null;
}

export interface BacktestStructuredResult {
  capital_label: string;
  rules: BacktestMetrics[];
  benchmarks: BacktestComparison[];
  limitations: string[];
}